src/mono/mono/metadata/marshal-ilgen.c

```diff
@@ -56,983 +57,33 @@ enum {
-		 * SafeHandle that was used during the marshalling,
-		 * for now we just ignore this, and ignore/discard any
-		 * changes that might have happened to the handle.
-		 */
-		break;
-	}
-
-	case MONO_MARSHAL_CONV_HANDLEREF: {
-		/*
-		 * Passing HandleRefs in a struct that is ref()ed does not
-		 * copy the values back to the HandleRef
-		 */
-		break;
-	}
-
-	case MONO_MARSHAL_CONV_ARRAY_SAVEARRAY:
-	default: {
-		char *msg = g_strdup_printf ("marshaling conversion %d not implemented", conv);
-
-		mono_mb_emit_exception_marshal_directive (mb, msg);
-		break;
-	}
-	}
-}
-
-// On legacy Mono, LPTSTR was either UTF16 or UTF8 depending on platform
-static inline MonoJitICallId
-mono_string_to_platform_unicode (void)
-{
-#ifdef TARGET_WIN32
-	return MONO_JIT_ICALL_mono_marshal_string_to_utf16;
-#else
-	return MONO_JIT_ICALL_mono_string_to_utf8str;
-#endif
-}
-
-static inline MonoJitICallId
-mono_string_from_platform_unicode (void)
-{
-#ifdef TARGET_WIN32
-	return MONO_JIT_ICALL_ves_icall_mono_string_from_utf16;
-#else
-	return MONO_JIT_ICALL_ves_icall_string_new_wrapper;
-#endif
-}
-
-static inline MonoJitICallId
-mono_string_builder_to_platform_unicode (void)
-{
-#ifdef TARGET_WIN32
-	return MONO_JIT_ICALL_mono_string_builder_to_utf16;
-#else
-	return MONO_JIT_ICALL_mono_string_builder_to_utf8;
-#endif
-}
-
-static inline MonoJitICallId
-mono_string_builder_from_platform_unicode (void)
-{
-#ifdef TARGET_WIN32
-	return MONO_JIT_ICALL_mono_string_utf16_to_builder;
-#else
-	return MONO_JIT_ICALL_mono_string_utf8_to_builder;
-#endif
-}
-
-static MonoMarshalConv
-conv_str_inverse (MonoMarshalConv conv)
-{
-	switch (conv) {
-	// AnsiBStr
-	case MONO_MARSHAL_CONV_STR_ANSIBSTR:
-		return MONO_MARSHAL_CONV_ANSIBSTR_STR;
-	case MONO_MARSHAL_CONV_ANSIBSTR_STR:
-		return MONO_MARSHAL_CONV_STR_ANSIBSTR;
-
-	// BStr
-	case MONO_MARSHAL_CONV_STR_BSTR:
-		return MONO_MARSHAL_CONV_BSTR_STR;
-	case MONO_MARSHAL_CONV_BSTR_STR:
-		return MONO_MARSHAL_CONV_STR_BSTR;
-
-	// LPStr
-	case MONO_MARSHAL_CONV_STR_LPSTR:
-		return MONO_MARSHAL_CONV_LPSTR_STR;
-	case MONO_MARSHAL_CONV_LPSTR_STR:
-		return MONO_MARSHAL_CONV_STR_LPSTR;
-
-	// LPTStr
-	case MONO_MARSHAL_CONV_STR_LPTSTR:
-		return MONO_MARSHAL_CONV_LPTSTR_STR;
-	case MONO_MARSHAL_CONV_LPTSTR_STR:
-		return MONO_MARSHAL_CONV_STR_LPTSTR;
-
-	// LPUTF8Str
-	case MONO_MARSHAL_CONV_STR_UTF8STR:
-		return MONO_MARSHAL_CONV_UTF8STR_STR;
-	case MONO_MARSHAL_CONV_UTF8STR_STR:
-		return MONO_MARSHAL_CONV_STR_UTF8STR;
-
-	// LPWStr
-	case MONO_MARSHAL_CONV_STR_LPWSTR:
-		return MONO_MARSHAL_CONV_LPWSTR_STR;
-	case MONO_MARSHAL_CONV_LPWSTR_STR:
-		return MONO_MARSHAL_CONV_STR_LPWSTR;
-
-	// TBStr
-	case MONO_MARSHAL_CONV_STR_TBSTR:
-		return MONO_MARSHAL_CONV_TBSTR_STR;
-	case MONO_MARSHAL_CONV_TBSTR_STR:
-		return MONO_MARSHAL_CONV_STR_TBSTR;
-
-	default:
-		g_assert_not_reached ();
-	}
-}
-
-static MonoJitICallId
-conv_to_icall (MonoMarshalConv conv, int *ind_store_type)
```

Comment:
i'm surprised this needs to be shared.

in the lightweight marshaler this should never get called (again because `conv` should never be anything other than `NONE`)

Reply:
I think, taking that into account I can probably remove it. But can I actually make that change in this PR? It seems like I need to seperate out the lightweight and ilgen first.

Reply:
I think this will probably have to wait until we have the lightweight marshaling code separated out.  right now I think conv_to_icall is getting called from both marshal-ilgen and marshal-shared, so it would be difficult to just have a single copy.  after the split, it will hopefully be easier to refactor the code so that -ilgen and -lightweight each get their own copy and the -ligthweight one is trivial.